sklearn/utils/validation.py

```diff
@@ -1986,8 +1986,11 @@ def _get_feature_names(X):
     feature_names = None
 
     # extract feature names for support array containers
-    if hasattr(X, "columns"):
+    if hasattr(X, "columns") and hasattr(X, "iloc"):
```

Comment:
we REALLY need a `_is_pandas_df` util 🙈

Reply:
I'm trying not to introduce a regression with pandas. Specifically `df.__dataframe__` with the default `allow_copy=True` can make copies.

The better approach is use the [DataFrame API SPEC](https://data-apis.org/dataframe-api/draft/API_specification/index.html). Concretely, [DataFrame.get_column_names](https://data-apis.org/dataframe-api/draft/API_specification/dataframe_object.html#dataframe_api.DataFrame.get_column_names). Currently, there is no library that supports the SPEC yet. 

There is an implementation of the DataFrame SPEC for pandas and polars: https://github.com/MarcoGorelli/impl-dataframe-api. The implementation for each library is one file each so it will not be hard for us to vendor and make use of the SPEC before the libraries adopt it.

As for this PR, I introduced a `_is_pandas_df` function that makes sure that `X` is a pandas DataFrame.

Reply:
+1 for temporarily vendoring the compat layer and try to only use the spec API in our estimators' code.

Reply:
The above can be done in a subsequent PR, no need to delay this one.